feat(rpc): add getunclaimedgas RPC method

ixje opened the pull request:
fix https://github.com/CityOfZion/neon-js/issues/593

snowypowers commented on packages/neon-core/src/rpc/Query.ts:
```diff
@@ -133,6 +133,11 @@ export interface ValidateAddressResult {
   isvalid: boolean;
 }
 
+export interface UnclaimedGasResult {
```
Lets get the name in line with the rest of the interfaces. Top level interfaces can follow the name of the RPC call to make the link very clear.
```suggestion
export interface GetUnclaimedGasResult{

```

snowypowers commented on packages/neon-core/src/rpc/Query.ts:
```diff
@@ -477,6 +482,15 @@ export class Query<TParams extends unknown[], TResponse> {
     });
   }
 
+  public static getUnclaimedGas(
```
A simple docstring describing the usage.

snowypowers commented on packages/neon-core/src/rpc/clients/NeoServerRpcClient.ts:
```diff
@@ -258,6 +258,14 @@ export function NeoServerRpcMixin<TBase extends RpcDispatcherMixin>(
+     */
+    public async getUnclaimedGas(addr: string): Promise<number> {
+      const response = await this.execute(Query.getUnclaimedGas(addr));
+      return parseInt(response.unclaimed) / 100_000_000;
```
This is possible unsafe if the initial parseInt is bigger than MAX_SAFE_INTEGER. Might also want to rethink about returning a number since that too might be unsafe but should be fine in this case.

```suggestion
      return BigInteger.fromNumber(response.unclaimed).div(100_000_000).toNumber();
```

(suggestion requires adding an import I think)

ixje replied:
I got a "property `toNumber()` does not exist on BigInteger". So instead I changed the return value to be BigInt. Which might be more suitable anyway, given that even after dividing by the decimals factor is might (although very unlikely) still exceed MAX_SAFE_INTEGER? Let me know what you think

snowypowers replied:
Wait I just realised we lose decimal precision in this case. We will need a better construct around this but for now Fixed8 will work. Look at Fixed8.fromRawNumber for your usage.

ixje replied:
Done

snowypowers reviewed (changes requested):
Needs a unit test to ensure the data parsing works.
Needs an integration test for smoke testing.

ixje commented:
> Needs a unit test to ensure the data parsing works.
> Needs an integration test for smoke testing.

Can you please point me in a direction. I'm still trying to find my way around TS and the `neon-js` library itself. Thanks

snowypowers commented:
neon-core/__integration__/clients/NeoServerRpcCLient.ts should be the place for integration test. We probably just wanna issue the call and ensure that we get a non-zero value (aka nothing blows up).

The unit tests is a little more tricky lets skip it for now.

snowypowers commented:
I have added a unit test with the mocks just to make sure that the parsing of the numbers work.

I have adjusted the integration test such that it will work with the integration container that we run against during normal tests. Your test will fail as the endpoint returns an error "invalid address".

ixje commented:
Thanks!

> On 10 Oct 2020, at 09:44, Yak Jun Xiang <notifications@github.com> wrote:
> 
> ﻿
> I have added a unit test with the mocks just to make sure that the parsing of the numbers work.
> 
> I have adjusted the integration test such that it will work with the integration container that we run against during normal tests. Your test will fail as the endpoint returns an error "invalid address".
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
